### PR TITLE
[lldb] Return index of element in ValueObject path instead of the element's value

### PR DESCRIPTION
**Pull request by PortalPete**
It's more meaningful and actionable to indicate which element in the array has an issue by returning that element's index instead of its value. The value can be ambiguous if at least one other element has the same value.

The first parameter for these methods is `idxs`, an array of indices that represent a path from a (root) parent to on of its descendants, typically though intermediate descendants. When the path leads to a descendant that doesn't exist, the method is supposed to indicate where things went wrong by setting an index to `&index_of_error`, the second parameter.

The problem is the method sets `*index_of_error` to the index of the most recent parent's child in the hierarchy, which isn't very useful if there's more one index with the same value in the path.

In this example, each element in the path has a value that's the same as another element.

```cpp
GetChildAtIndexPath({1, 2, 3, 3, 1, 1, 2}, &index_of_error);
```

Say the the second `1` in the path (the 5th element at `[4]`) doesn't exist and the code returns a `nullptr`. In that situation, the code sets `*index_of_error` to `1`, but that's an ambiguous hint can implicate the 1st, 5th, or 6th element (at `[0]`, `[4]`, or `[5]`).

It’s more helpful to set `*index_of_error` to `4` to clearly indicate which element in `idxs` has the issue.

### CODE REVIEW

**Comment by DavidSpickett** [comment]
I can see why the mistake was made, we have an index into a list of different indexes? Anyway, please add a test for this.

File: lldb/source/Core/ValueObject.cpp
Comment (chelcassanova) [inline lldb/source/Core/ValueObject.cpp:410]: This is the tiniest nit on earth but `current_index++` here to increment this?

**Comment by felipepiovezan** [comment]
Should we change the documentation of these functions?

File: lldb/source/Core/ValueObject.cpp
Comment (PortalPete) [reply lldb/source/Core/ValueObject.cpp:410]: Sure!
I had a hunch someone would point this one out, but I thought it'd be fun to see who, if anyone, would.
You passed the test! 😂 

Personally, I prefer `+= 1` because:
- Swift doesn't have the `++` operator.
- In my mind `+= 1` more explicitly communicates my intent.
- I've seen people get burned by the "post" part of a post-increment operator too many times. 🙃

File: lldb/source/Core/ValueObject.cpp
Comment (PortalPete) [reply lldb/source/Core/ValueObject.cpp:430]: I'm actually going to go with `++current_index` because that's what [the LLVM coding style](https://llvm.org/docs/CodingStandards.html#prefer-preincrement) recommends.


> Hard fast rule: Preincrement (++X) may be no slower than postincrement (X++) and could very well be a lot faster than it. Use preincrementation whenever possible.
>
> ...

File: lldb/source/Core/ValueObject.cpp
Comment (chelcassanova) [reply lldb/source/Core/ValueObject.cpp:410]: After reading the Swift rationale for removing the `++` operator I understand why you'd use `+= 1` instead.

File: lldb/source/Core/ValueObject.cpp
Comment (felipepiovezan) [reply lldb/source/Core/ValueObject.cpp:430]: Note this guideline is about pre vs post, and it refers to iterators mostly.

I don't think you need to change anything here!

File: lldb/source/Core/ValueObject.cpp
Comment (PortalPete) [reply lldb/source/Core/ValueObject.cpp]: I take it you're referring to [Swift evolution's proposal 0004 - Remove the `++` and `--` operators](https://github.com/apple/swift-evolution/blob/main/proposals/0004-remove-pre-post-inc-decrement.md).

As a coding and computer science mentor and writer, I naturally agree with Chris Lattner in that any code I write may serves as potential educational material to the reader/maintainer. For me, there are more cons than pros to the pre- and post-increment operators when I look at them through an educational lens.

File: lldb/source/Core/ValueObject.cpp
Comment (PortalPete) [reply lldb/source/Core/ValueObject.cpp]: Right, but if we're not going with `+= 1` (for consistency and/or informal style?), then I prefer to use the pre-increment operator.

Or are you saying we should just stick with `+= 1` now?

**Comment by PortalPete** [comment]
> I can see why the mistake was made, we have an index into a list of different indexes? Anyway, please add a test for this.

I'm not really sure there's a good way to test this there doesn't appear to be anything that calls these methods, let alone an SB API.

CC @adrian-prantl

File: lldb/source/Core/ValueObject.cpp
Comment (PortalPete) [reply lldb/source/Core/ValueObject.cpp]: Ok, I'll just stick with `+= 1` and hope it becomes a Swift-inspired trend.

**Comment by adrian-prantl** [comment]
> > doesn't appear to be anything that calls these methods
> 
> Candidate for removal then? If so it still makes sense to fix them, then remove them. Saves fixing them again if they are brought back.
> 
> Either way, I'm not going to block this by requiring a test case.

I think that makes sense. @PortalPete why don't you land this without a test and then create a PR to remove the unused methods?

**Comment by PortalPete** [comment]
> > > doesn't appear to be anything that calls these methods
> > 
> > 
> > Candidate for removal then? If so it still makes sense to fix them, then remove them. Saves fixing them again if they are brought back.
> > Either way, I'm not going to block this by requiring a test case.
> 
> I think that makes sense. @PortalPete why don't you land this without a test and then create a PR to remove the unused methods?

That works for me, @adrian-prantl.
I'll remove these methods wholesale after someone approves/merges this PR on my behalf.

**Review by felipepiovezan — APPROVED** [review]

**Pull request merged**
